Implements SampledValue

gvegayon opened the pull request:
- Introduces the `SampledValye` class with `value`, `t_start`, and `t_unit` attributes (time-related attributes are optional).
- Updates all `RandomVariable.sample()` calls to return tuples/namedtuples with `SampledValue` instead of `ArrayLike` objects.
- Tweaks one of the tutorials so that the figures show reasonable values.

![image](https://github.com/user-attachments/assets/0264a182-6ab5-4b4d-b70e-2496915bd26d)

The latter is the new version of the last plot, the previous figure looks like the following:

![image](https://github.com/user-attachments/assets/0a841340-fbbc-4cba-8ba2-873ac780672e)

~~- It fixes a hidden bug in the hospital admissions model tutorial where the day-of-the-week effect was not properly used.~~


## For later

Also, in the extending pyrenew tutorial



![image](https://github.com/user-attachments/assets/9ecce1ec-eb7e-4c0d-af23-87361ba30de7)
New version of the hospital admissions plot. vs:
![image](https://github.com/user-attachments/assets/c5495bbf-cc07-4789-bf3b-f9ecd62eb3a9)


**This is not implemented as the section on using the day-of-week effect was removed from the tutorial.**

gvegayon commented:
Just found a bug in the code. Particularly the definition of the day of the week effect in the `hospital_admissions_model.qmd` tutorial. The current implementation does not use all the day-of-the-week effect samples but rather the first observation (Monday/Sunday, depending on where you live). Here is the code:

https://github.com/CDCgov/multisignal-epi-inference/blob/ce055afcaa2998d152ce39b3efd248d98eb3ddef/docs/source/tutorials/hospital_admissions_model.qmd#L502

In other words, instead of returning a tuple/namedtuple, it returns a jax array, which then, when used within `latent.HospitalAdmissions.sample()` fails to use the full vector:

https://github.com/CDCgov/multisignal-epi-inference/blob/ce055afcaa2998d152ce39b3efd248d98eb3ddef/model/src/pyrenew/latent/hospitaladmissions.py#L192-L195

The example should instead return a tuple:

```python
         return (jnp.tile(ans, self.nweeks)[: self.len],)
```

But then, the code fails because the day-of-the-week RV samples a vector of length `dat.shape[0] = 90` and tries multiplying it by a vector of infections of length `dat.shape[0] + padding (21) + seeding (15) = 126`.

For this PR, I'll fix the example by returning a tuple with a `TimeArray` (as all RVs in the PR), and specifying the sample size to be 126. **A better solution will be to refactor `latent.HospitalAdmissions` to use a `TimeArray` instead of a `ArrayLike` for the day-of-the-week effect, passing the `latent_admissions` as a `TimeArray` so it will figure out what is the proper length (and alignment) automatically.**

This could have been identified if we checked that all calls to `RandomVariable.sample()` return a namedtuple or a tuple of `TimeArray`s. I suggest leveraging we now have the `RandomVariable.__call__()` method to introduce a check there and have `RandomVariable.sample()` be renamed to `RandomVariable._sample_()` so it is now considered to be internal.

attn @damonbayer @dylanhmorris

gvegayon commented on docs/source/tutorials/hospital_admissions_model.qmd:
```diff
@@ -499,14 +490,16 @@ class DayOfWeekEffect(metaclass.RandomVariable):
 
-        return jnp.tile(ans, self.nweeks)[: self.len]
+        return (metaclass.TimeArray(
+            jnp.tile(ans, 100)[: n_timepoints]),
```
I haven't been able to figure out why using `int(jnp.ceil(n_timepoints/7))` instead of hardcoding 100 doesn't work. I understand `jax` usually doesn't like that pattern, but the periodic effect works (see [here](https://github.com/CDCgov/multisignal-epi-inference/blob/ce055afcaa2998d152ce39b3efd248d98eb3ddef/model/src/pyrenew/process/periodiceffect.py#L64-L68)). Pre-specifying the number of weeks this effect needs to be tiled beforehand is not an option as it won't work with the posterior predictive method.

damonbayer replied:
In what way doesn't it work? Do you get an error?

dylanhmorris replied:
I'd suggest staying away from this pattern entirely. I'd have the returned array be of length 7 and then use a mask rather than a tiling to create an array aligned with the actual data. In other words, if I have some vector of dow (here a scaled simplex):
```python
dow_effects = 7 * jnp.array([0.1, 0.2, 0.1, 0.2, 0.1, 0.2, 0.1])
```

And I have a vector of values to which I'd like to associate the d.o.w. effect, which might be more or fewer than 7 entries long:

```python
pre_dow_effect_values_shorter = jnp.array([15, 7, 2, 4, 6])
pre_dow_effect_values_longer = jnp.array([35, 32, 13, 5, 10, 15, 36, 20, 3])
```

I do this with a lookup mask, e.g.:

```python
dow_index_shorter = jnp.array([5, 6, 0, 1, 2])
dow_index_longer = jnp.array([3, 4, 5, 6, 0, 1, 2, 3, 4])
```

I can then do:

```python
dow_effects_shorter = dow_effects[dow_index_shorter]
dow_effects_longer = dow_effects[dow_index_longer]
```

gvegayon replied:
> In what way doesn't it work? Do you get an error?

That's right, @damonbayer. The compiler doesn't like casting floats into ints in some situations. Not always the case, as the periodic effects work just fine. Will work on this next week.

damonbayer replied:
> > In what way doesn't it work? Do you get an error?
> 
> That's right, @damonbayer. The compiler doesn't like casting floats into ints in some situations. Not always the case, as the periodic effects work just fine. Will work on this next week.

Did you try integer division? https://stackoverflow.com/a/17511341

gvegayon replied:
I'll try that, @damonbayer. Thanks!

gvegayon replied:
Works like a charm, @damonbayer! Thanks.

damonbayer commented on model/src/pyrenew/metaclass.py:
```diff
@@ -152,6 +192,15 @@ def set_timeseries(
+        # Either both values are None or both are not None
+        assert (t_unit is not None and t_start is not None) or (
+            t_unit is None and t_start is None
+        ), "Both t_start and t_unit should be None or not None."
```
I think we could be more helpful here by printing `t_start` and `t_unit` in the error message.

damonbayer commented:
@gvegayon did you consider implementing this with `__jax_array__`?

gvegayon commented:
> @gvegayon did you consider implementing this with `__jax_array__`? I think you did, but I couldn’t remember.

I have not. I was checking, and there's not enough documentation to make good use of it. Looking at the source code of `jax`, it seems we would have to implement `__jax_array__` as a function call. I suggest adding this as an issue and moving forward; mostly because it is not properly documented yet.

gvegayon commented:
The current conflict includes the deleted component on the day-of-the-week effects in https://github.com/CDCgov/multisignal-epi-inference/pull/281. This PR fixes it; should I resolve the conflict by including the previously deleted effect?

dylanhmorris commented on model/src/pyrenew/deterministic/deterministic.py:
I could be persuaded, but I do not think `DeterministicVariable`s should necessarily be `TimeArray`s.

If we want a deterministic, time-aware RV, wouldn't it be simpler to do it this way?

```python
my_time_aware_value = TimeArray([5, 23, 4], t_start=-4, t_unit=1)
my_time_aware_deterministic = DeterministicVariable(my_time_aware_value)
```

gvegayon replied:
> > The current conflict includes the deleted component on the day-of-the-week effects in #281. This PR fixes it; should I resolve the conflict by including the previously deleted effect?
> 
> Could we demonstrate it in the "Time handling in pyrenew" tutorial instead?

Surely, but I would like to do that in the next PR then. I think this would be a good opportunity to make use of the `TimeArray` attributes, for which we would need to: (a) encode time in the observed data, which we are not doing right now, and (b) refactor the implementation of the latent hospitalizations module so it would pass that information to the day-of-the-week effect.

dylanhmorris replied:
We previously discussed having a `SampledValue` class whose `value` attribute can be an array or something else and whose `t_start` and `t_unit` attributes can be `None`. For example, both of the below approaches represent a time-unaware sampled scalar, but I think the first is clearer about the intent:

```python
var = SampledValue(value=23.523)
var.value
```

versus:
```python
var = TimeArray(array=23.532)
var.array
```

dylanhmorris replied:
One could even image having a time-unaware base `SampledValue` class, and adding the `t_start` and `t_unit` attributes only in some subclass `TimeAwareSampledValue` (versus just having `SampledValue`s have the option of `t_start = None` and `t_unit = None`). But that might add extra hierarchy/complexity with minimal benefit

gvegayon replied:
All now is SampledValue

gvegayon commented:
> > The current conflict includes the deleted component on the day-of-the-week effects in #281. This PR fixes it; should I resolve the conflict by including the previously deleted effect?
> 
> Could we demonstrate it in the "Time handling in pyrenew" tutorial instead?

Surely, but I would like to do that in the next PR then. I think this would be a good opportunity to make use of the `TimeArray` attributes, for which we would need to: (a) encode time in the observed data, which we are not doing right now, and (b) refactor the implementation of the latent hospitalizations module so it would pass that information to the day-of-the-week effect.

gvegayon commented:
> > The current conflict includes the deleted component on the day-of-the-week effects in #281. This PR fixes it; should I resolve the conflict by including the previously deleted effect?
> 
> Could we demonstrate it in the "Time handling in pyrenew" tutorial instead?

Surely, but I would like to do that in the next PR then. I think this would be a good opportunity to make use of the `TimeArray` attributes, for which we would need to: (a) encode time in the observed data, which we are not doing right now, and (b) refactor the implementation of the latent hospitalizations module so it would pass that information to the day-of-the-week effect.

(sorry I added and deleted this comment)

gvegayon commented:
Ready for review, @damonbayer and @dylanhmorris

gvegayon commented:
Fixed the latest conflicts, @damonbayer and @dylanhmorris.

damonbayer commented on model/src/pyrenew/process/periodiceffect.py:
```diff
@@ -110,9 +114,11 @@ def sample(self, duration: int, **kwargs):
-            value=self.broadcaster(
-                data=self.quantity_to_broadcast.sample(**kwargs)[0],
-                n_timepoints=duration,
+            value=SampledValue(
```
Should this return a `SampledValue`, not an `PeriodicEffectSample`?

damonbayer replied:
Actually, I think the plan was to return a regular tuple, and later have a PR to convert everything to named tuples.

gvegayon replied:
PeriodicEffectSample is (was) a namedtuple before this PR. So there's no need to change here.

damonbayer replied:
I would prefer to remove `PeriodicEffectSample` and just return a namedtuple without a special class.

damonbayer commented on model/src/pyrenew/process/rtperiodicdiff.py:
```diff
@@ -188,7 +192,9 @@ def sample(
 
         return RtPeriodicDiffProcessSample(
-            rt=self.broadcaster(jnp.exp(log_rt.flatten()), duration),
+            rt=SampledValue(
```
Same as before, should we remove the extra layer of abstraction and just return a named tuple with a `SampledValue`?

gvegayon replied:
Same as my last comment, nothing to do here.

damonbayer commented on model/src/pyrenew/model/admissionsmodel.py:
I think the types in `HospModelSample(NamedTuple)` are now incorrect. We are returning `SampledValue`s instead of `ArrayLike`s.

damonbayer commented on model/src/pyrenew/model/rtinfectionsrenewalmodel.py:
Update types in `class RtInfectionsRenewalSample(NamedTuple)`

gvegayon replied:
Noted

damonbayer commented on model/src/pyrenew/metaclass.py:
```diff
@@ -94,6 +94,46 @@ def _assert_sample_and_rtype(
+        value : ArrayLike
+            The sampled value.
+        t_start : int, optional
+            The start time of the data.
```
```suggestion
            The start time of the value.
```

damonbayer commented on model/src/pyrenew/latent/infectionswithfeedback.py:
Update types in `class InfectionsRtFeedbackSample(NamedTuple)`

gvegayon replied:
Noted

damonbayer commented on docs/source/tutorials/time.qmd:
```diff
@@ -10,7 +10,7 @@ The fundamental time unit should represent a period of fixed (or approximately f
 For many infectious disease renewal models of interest, the fundamental time unit will be days, and we will proceed with this tutorial treating days as our fundamental unit.
 
- `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units.
+ `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Moreover, return values from `RandomVariable.sample()` are namedtuples with `SampledValue` objects that carry the same information.
```
This is not currently true. We don't always return NameedTuples

gvegayon replied:
True. Will correct.

damonbayer reviewed (changes requested):
Just a few things to clean up.

dylanhmorris commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -83,8 +90,14 @@ def sample(
         -------
         tuple
-            Containing the stored values during construction.
+            Containing the stored values during construction wrapped in a SampledValue.
```
```suggestion
            A length-one tuple whose single entry is a :class:`SampledValue` 
            instance with `value=self.vars`, `t_start=self.t_start`, and `t_unit=self.t_unit`.
```

dylanhmorris commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -28,13 +30,18 @@ def __init__(
             A tuple with arraylike objects.
```
```suggestion
            An arraylike object.
```

dylanhmorris replied:
We should discuss elsewhere whether `DeterministicVariable`s should be able to have an arbitrary `pytree` as their `vars` value. I lean yes.

I would also support renaming `vars` to `value`, for consistency with `SampledValue`.

For clarity, both of the above are out of scope for this PR.

gvegayon replied:
Issue created

dylanhmorris commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -83,8 +90,14 @@ def sample(
         if record:
             npro.deterministic(self.name, self.vars)
-        return (self.vars,)
+        return (
+            SampledValue(
+                value=self.vars,
+                t_start=self.t_start,
+                t_unit=self.t_unit,
+            ),
+        )
```
Noting here that this will be the gotcha of the new system. If I have 

```python
some_time_aware_sampled_value = stochastic_rv.sample()
```
and I want to pass it somewhere else downstream as a `DeterministicVariable`, I should _not_ do:

```python
my_det_var = DeterministicVariable(vars=some_time_aware_sampled_value,
                                                             name="my_det_var")
```

Instead one should do:

```python
my_det_var = DeterministicVariable(
               vars=some_time_aware_sampled_value.value,
               name="my_det_var", 
               t_start=some_time_aware_sampled_value.t_start,
               t_unit=some_time_aware_sampled_value.t_unit)
```

dylanhmorris replied:
Note that if we do #307 and also take my suggestion above of renaming the `vars` attribute of `DeterministicVariable` to `value`, then we could use unpacking here to turn `SampledValues` into `DeterministicVariables`:

```python
my_det_var = DeterministicVariable(
               name="my_det_var", 
              **some_time_aware_sampled_value)
```
I am on the fence RE: whether this is nicely succinct or too opaque/magic.

dylanhmorris replied:
In any case we should document the gotcha somewhere, perhaps.

gvegayon replied:
The current implementation checks that `vars` are indeed `ArrayLike`, which the `SampledValue` should not be, so I don't think this will cause such issues.

dylanhmorris replied:
I was thinking more about usage guidance than erroring, though the error is nice to have.

dylanhmorris replied:
Sounds good. Flag as todo in #304?

dylanhmorris commented on model/src/pyrenew/deterministic/deterministicpmf.py:
```diff
@@ -18,6 +18,8 @@ def __init__(
         vars: ArrayLike,
         name: str,
         tol: float = 1e-5,
```
Out of scope here but this should presumably have a default set higher up. See also #268

dylanhmorris commented on model/src/pyrenew/deterministic/deterministicpmf.py:
```diff
@@ -46,7 +53,12 @@ def __init__(
+            name=name,
+            t_start=t_start,
+            t_unit=t_unit,
+        )
```
Beyond the scope of this PR but I think this is most efficiently handled by actual inheritance, e.g. something like:

```python
def __init__(
       self, 
       name: str
       pmf: ArrayLike,
       tol: float = 1e-5,
       **kwargs):
    normed_pmf = validate_discrete_dist_vector(pmf, tol)
    super().__init__(vars=normed_pmf, **kwargs)
```
This avoids having to override the `sample()` method, among other things.

gvegayon replied:
Smells like a wider discussion, if we consider the current preference of using composition instead of inheritance when building models.

dylanhmorris commented on model/src/pyrenew/deterministic/nullrv.py:
```diff
@@ -46,10 +47,10 @@ def sample(
+        return (SampledValue(None),)
 
 
 class NullProcess(NullVariable):
```
Is `NullProcess` still used? It seems to be completely identical to `NullVariable`. Artifact of when there were `RandomProcess`es that were separate/distinct from `RandomVariable`s?

gvegayon replied:
> Is `NullProcess` still used? It seems to be completely identical to `NullVariable`. Artifact of when there were `RandomProcess`es that were separate/distinct from `RandomVariable`s?

Unsure if this is used right now. The main difference is that `process.*` has the `duration` argument. At some point, the idea of making process, latent, observed, etc. different animals came up. This could add to it.

dylanhmorris commented on model/src/pyrenew/deterministic/nullrv.py:
```diff
@@ -95,10 +96,10 @@ def sample(
+        return (SampledValue(None),)
 
 
 class NullObservation(NullVariable):
```
Similarly, is this still used?

gvegayon replied:
Same as before.

dylanhmorris commented on model/src/pyrenew/deterministic/process.py:
```diff
@@ -2,6 +2,7 @@
+from pyrenew.metaclass import SampledValue
 
 
 class DeterministicProcess(DeterministicVariable):
```
I'd also suggest reviewing the implementation here. I think given some other choices we've made we'll perhaps want to modify it. Out of scope for this PR, however.

gvegayon replied:
Agreed

dylanhmorris commented on model/src/pyrenew/deterministic/process.py:
```diff
@@ -29,14 +30,26 @@ def sample(
-            return (jnp.hstack([res, jnp.repeat(res[-1], dif)]),)
-
-        return (res[:duration],)
+            return (
+                SampledValue(
+                    jnp.hstack([res.value, jnp.repeat(res.value[-1], dif)]),
+                    t_start=self.t_start,
+                    t_unit=self.t_unit,
+                ),
+            )
+
+        return (
+            SampledValue(
+                value=res.value[:duration],
+                t_start=self.t_start,
+                t_unit=self.t_unit,
+            ),
+        )
```
Double return here. Typo?

gvegayon replied:
I don't think so. Both return are different.

dylanhmorris replied:
Oh I see one is inside of an `if`. For clarity, I'd suggest this pattern

```python
if <condition>:
      result = <value>
else:
     result = <other value>
return result
```

dylanhmorris commented on model/src/pyrenew/latent/hospitaladmissions.py:
```diff
@@ -162,7 +162,7 @@ def sample(
         Parameters
         ----------
-        latent : ArrayLike
+        latent : ArrayLike or SampledValue
```
```suggestion
        latent_infections : ArrayLike
```
As the code written, this cannot be a `SampledValue`, since we do not call `.value` on it. And I think that's at least arguably correct

gvegayon replied:
Good catch.

dylanhmorris commented on model/src/pyrenew/latent/hospitaladmissions.py:
```diff
@@ -184,25 +184,33 @@ def sample(
+            infection_hosp_rate=infection_hosp_rate,
+            latent_hospital_admissions=SampledValue(
+                value=latent_hospital_admissions,
+                t_start=self.infection_to_admission_interval_rv.t_start,
+                t_unit=self.infection_to_admission_interval_rv.t_unit,
```
I don't think these are the relevant t_start/t_unit.

I would have thought they should be the same as the RV itself (or, eventually computed dynamically from the values in a time-aware `latent_infections` that gets passed + attributes of the delay interval, if we eventually go in that direction). For now, suggest:

```suggestion
                t_start=self.t_start,
                t_unit=self.t_unit,
```

dylanhmorris commented on model/src/pyrenew/latent/infection_initialization_method.py:
```diff
@@ -177,6 +177,7 @@ def initialize_infections(self, I_pre_init: ArrayLike):
                 f"I_pre_init must be an array of size 1. Got size {I_pre_init.size}."
             )
         (rate,) = self.rate()
+        rate = rate.value
```
```suggestion
        rate = self.rate()[0].value
```

dylanhmorris replied:
Longer term this should be refactored to follow the [variable naming conventions](https://github.com/CDCgov/multisignal-epi-inference/blob/main/docs/source/developer_documentation.rst#pyrenew-principles), which would allow us to do something like:

```python
rate = self.rate_rv()[0].value
```

dylanhmorris commented on model/src/pyrenew/latent/infectionswithfeedback.py:
```diff
@@ -159,6 +163,7 @@ def sample(
         inf_feedback_strength, *_ = self.infection_feedback_strength(
```
similar comment to sampling `rate` above.

dylanhmorris commented on model/src/pyrenew/metaclass.py:
```diff
@@ -94,6 +94,46 @@ def _assert_sample_and_rtype(
+        None
+        """
+
+        if value is not None:
+            assert isinstance(
+                value, ArrayLike
+            ), "value should be an array-like object."
```
Are we sure we want this (versus arbitrary `pytrees`?

gvegayon replied:
Good question. I would say yes, keep things as ArrayLike objects, but pytrees should be, in principle, OK. Since I'm making the SampledValue a namedtuple, I'm taking this out and we can leave it for a later discussion if we want to enforce a type.

dylanhmorris commented on model/src/pyrenew/metaclass.py:
```diff
@@ -94,6 +94,46 @@ def _assert_sample_and_rtype(
     return None
 
 
+class SampledValue:
```
I thought the plan was to have `SampledValue`s be `NamedTuple`s to ensure that they are immutable.

dylanhmorris reviewed (changes requested):
Stopping a bit of the way through because I have a fairly major question. Here, `SampledValue` is a regular class. I had thought the plan was to have `SampledValue`s be NamedTuples to ensure that they are immutable.

gvegayon commented on docs/source/tutorials/time.qmd:
```diff
@@ -10,7 +10,7 @@ The fundamental time unit should represent a period of fixed (or approximately f
 For many infectious disease renewal models of interest, the fundamental time unit will be days, and we will proceed with this tutorial treating days as our fundamental unit.
 
- `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units.
+ `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Moreover, return values from `RandomVariable.sample()` are namedtuples with `SampledValue` objects that carry the same information.
```
```suggestion
 `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Return values from `RandomVariable.sample()` are `tuples` or `namedtuple`s of `SampledValue` objects. Each such `SampledValue` is optionally time-aware with specifiable `t_start` and `t_unit` attributes.
```

damonbayer commented on model/src/test/test_infectionsrtfeedback.py:
```diff
@@ -157,3 +157,5 @@ def test_infectionsrtfeedback_feedback():
 
     return None
+
+test_infectionsrtfeedback()
```
```suggestion

```

gvegayon replied:
Yes, also noted that. Solved now.

dylanhmorris commented on model/src/pyrenew/observation/negativebinomial.py:
```diff
@@ -93,8 +93,9 @@ def sample(
         )
-        return (negative_binomial_sample,)
+
+        return (SampledValue(negative_binomial_sample),)
```
I think we should probably pass `self.t_start` and `self.t_unit` here, even if they're typically going to be `None`.

gvegayon replied:
Noted

dylanhmorris commented on model/src/pyrenew/observation/poisson.py:
```diff
@@ -72,4 +72,4 @@ def sample(
             obs=obs,
         )
-        return (poisson_sample,)
+        return (SampledValue(poisson_sample),)
```
As above, I think we should probably pass `self.t_start` and `self.t_unit` here, even if they're typically going to be `None`.

gvegayon replied:
Noted

dylanhmorris commented on model/src/pyrenew/process/ar.py:
```diff
@@ -91,7 +91,7 @@ def _ar_scanner(carry, next):  # numpydoc ignore=GL08
 
         last, ts = lax.scan(_ar_scanner, inits - self.mean, noise)
-        return (jnp.hstack([inits, self.mean + ts.flatten()]),)
+        return (SampledValue(jnp.hstack([inits, self.mean + ts.flatten()])),)
```
As above, I think we should probably pass `self.t_start` and `self.t_unit` here, even if they're typically going to be `None`.

gvegayon replied:
Noted

dylanhmorris commented on model/src/pyrenew/process/firstdifferencear.py:
```diff
@@ -72,7 +72,11 @@ def sample(
             name=name + "_rate_of_change",
         )
-        return (init_val + jnp.cumsum(rates_of_change.flatten()),)
+        return (
+            SampledValue(
+                init_val + jnp.cumsum(rates_of_change.value.flatten())
+            ),
+        )
```
As above, I think we should probably pass `self.t_start` and `self.t_unit` here, even if they're typically going to be `None`.

dylanhmorris commented on model/src/pyrenew/process/periodiceffect.py:
```diff
@@ -110,9 +113,11 @@ def sample(self, duration: int, **kwargs):
-                data=self.quantity_to_broadcast.sample(**kwargs)[0],
-                n_timepoints=duration,
+            value=SampledValue(
+                self.broadcaster(
+                    data=self.quantity_to_broadcast.sample(**kwargs)[0].value,
+                    n_timepoints=duration,
```
We definitely want these SampledValues be time-aware, I'd think. For now I'd use `self.t_start` and `self.t_unit`, but we may want functionality to optionally compute those at construction from a time-aware `quantity_to_broadcast`

gvegayon replied:
Created an issue to list all components that will need to be explicitly timeaware: #320. Added `PeriodicEffect` to the list. 

att @damonbayer

dylanhmorris commented on model/src/pyrenew/process/rtperiodicdiff.py:
```diff
@@ -188,7 +192,9 @@ def sample(
-            rt=self.broadcaster(jnp.exp(log_rt.flatten()), duration),
+            rt=SampledValue(
+                self.broadcaster(jnp.exp(log_rt.value.flatten()), duration)
+            ),
```
As above, I think we should probably pass `self.t_start` and `self.t_unit` here.

gvegayon replied:
Noted

dylanhmorris commented on model/src/pyrenew/process/simplerandomwalk.py:
```diff
@@ -77,16 +77,16 @@ def sample(
         )
 
-        return (jnp.hstack([init, x.flatten()]),)
+        return (SampledValue(jnp.hstack([init.value, x.flatten()])),)
```
As above, I think we should probably pass `self.t_start` and `self.t_unit` here, even if they're sometimes going to be `None`.

gvegayon replied:
Noted

dylanhmorris commented on model/src/test/test_periodiceffect.py:
```diff
@@ -44,7 +44,7 @@ def test_periodiceffect() -> None:
     pe = PeriodicEffect(**params)
     with npro.handlers.seed(rng_seed=np.random.randint(1, 600)):
-        ans2 = pe(duration=duration).value
+        ans2 = pe(duration=duration).value.value
```
I think this has implications for our naming conventions for the `NamedTuples` returned by `RandomVariable.sample()`. For now I would suggest this:
```suggestion
        ans2 = pe(duration=duration)[0].value
```

dylanhmorris commented on model/src/test/test_periodiceffect.py:
```diff
@@ -81,9 +81,13 @@ def test_weeklyeffect() -> None:
-    ans1 = pe(duration=duration).value
-    ans2 = pe2(duration=duration).value
+    ans1 = pe(duration=duration).value.value
+    ans2 = pe2(duration=duration).value.value
```
As above

gvegayon replied:
Noted

dylanhmorris commented on model/src/test/test_periodiceffect.py:
```diff
@@ -81,9 +81,13 @@ def test_weeklyeffect() -> None:
     return None
+
+
+test_periodiceffect()
+test_weeklyeffect()
```
```suggestion
```

dylanhmorris reviewed (changes requested):
I think we almost always want `RandomVariables` to pass their time information (if any) forward in `SampledValue`s (default unmodified; modified in specific cases where it makes sense)

gvegayon commented:
It should be good to go now.

dylanhmorris commented:
Looks like some pre-commit and test fixes still needed @gvegayon

dylanhmorris commented on model/src/pyrenew/latent/infectionswithfeedback.py:
```diff
@@ -195,6 +199,8 @@ def sample(
         return InfectionsRtFeedbackSample(
-            post_initialization_infections=post_initialization_infections,
-            rt=Rt_adj,
+            post_initialization_infections=SampledValue(
+                post_initialization_infections
```
```suggestion
            post_initialization_infections=SampledValue(
                value=post_initialization_infections,
                t_start=self.t_start,
                t_unit=self.t_unit)
```

dylanhmorris reviewed (changes requested):
A few remaining tweaks

dylanhmorris commented:
This should be ready for review @damonbayer. @AFg6K7h4fhy2 I would like to note that I introduced a typo while resolving merge conflicts and the typos hook caught it, so thanks again for setting that up/troubleshooting it.

damonbayer commented on model/src/pyrenew/metaclass.py:
```diff
@@ -94,6 +94,28 @@ def _assert_sample_and_rtype(
+
+    Attributes
+    ----------
+    value : ArrayLike, optional
```
When would `value` be optional?

damonbayer replied:
Answering my own question, in `model/src/pyrenew/deterministic/nullrv.py`.

damonbayer commented on docs/source/tutorials/time.qmd:
```diff
@@ -10,9 +10,9 @@ The fundamental time unit should represent a period of fixed (or approximately f
 For many infectious disease renewal models of interest, the fundamental time unit will be days, and we will proceed with this tutorial treating days as our fundamental unit.
 
- `pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units.
+`pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Return values from `RandomVariable.sample()` are `tuples` or `namedtuple`s of `SampledValue` objects. Each such `SampledValue` is optiona ...
```
`RandomVariable`s having their own `t_start` and `t_unit` seems to contrast with the idea the `RandomVariable.sample()` can return multiple `SampledValue`s with different `t_start` and `t_unit`.

dylanhmorris replied:
Attempted clarification in 859dbc9: 

https://github.com/CDCgov/multisignal-epi-inference/blob/859dbc96712a887bf2a5daa98156bebbedbc8dab/docs/source/tutorials/time.qmd#L13-L15

dylanhmorris replied:
In the longer run, we should add an example of the second thing to the tutorial, but I think that can wait for the next time-handling PR.

damonbayer commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -75,16 +81,27 @@ def sample(
         ----------
         record : bool, optional
-            Whether to record the value of the deterministic RandomVariable. Defaults to True.
+            Whether to record the value of the deterministic
```
Why does this line change? Can we check this type of formatting with precommit?

dylanhmorris replied:
That was me disliking the long line. But yes, we should be able to set a maximum docstring line length etc

damonbayer commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -75,16 +81,27 @@ def sample(
+        tuple[SampledValue]
+            A length-one tuple whose single entry is a
+            :class:`SampledValue`
+            instance with `value=self.vars`,
```
I think we no longer use `vars`.

dylanhmorris replied:
Good catch, thanks. Fixed in b298bc2

damonbayer commented on model/src/pyrenew/deterministic/process.py:
```diff
@@ -29,14 +30,28 @@ def sample(
         -------
         tuple
-            Containing the stored values during construction.
+            Containing the stored values during construction wrapped in a SampledValue.
```
This doesn't read well to me. Not sure what it is trying to say.

dylanhmorris replied:
Made an attempt in 9ff9dea but I increasingly think we should get rid of the `DeterministicProcess` class, as I think it should be replaced by other more time-aware ways of handling temporal broadcasting of values. Beyond the scope of this PR, though.

damonbayer commented on model/src/pyrenew/deterministic/process.py:
```diff
@@ -29,14 +30,28 @@ def sample(
-            return (jnp.hstack([res, jnp.repeat(res[-1], dif)]),)
-
-        return (res[:duration],)
+            res = (
+                SampledValue(
+                    jnp.hstack([res.value, jnp.repeat(res.value[-1], dif)]),
+                    t_start=self.t_start,
+                    t_unit=self.t_unit,
+                ),
+            )
+        else:
+            res = (
+                SampledValue(
+                    value=res.value[:duration],
+                    t_start=self.t_start,
+                    t_unit=self.t_unit,
+                ),
+            )
+
+        return res
```
```suggestion
            value = jnp.hstack([res.value, jnp.repeat(res.value[-1], dif)])
        else:
            value = res.value[:duration]

        res = SampledValue(
            value,
            t_start=self.t_start,
            t_unit=self.t_unit,
        )

        return (res,)
```

dylanhmorris replied:
Good edit, thanks!

damonbayer commented on model/src/pyrenew/latent/hospitaladmissions.py:
```diff
@@ -20,12 +20,12 @@ class HospitalAdmissionsSample(NamedTuple):
     ----------
     infection_hosp_rate : float, optional
```
should this be a `SampledValue`?

dylanhmorris replied:
Yes it should. Fixed in 2177215

damonbayer commented on model/src/pyrenew/process/firstdifferencear.py:
```diff
@@ -75,7 +75,13 @@ def sample(
-        return (init_val + jnp.cumsum(rates_of_change.flatten()),)
+        return (
+            SampledValue(
+                init_val + jnp.cumsum(rates_of_change.value.flatten()),
```
Is the `.flatten` neccessar?

dylanhmorris replied:
Can investigate in #328

dylanhmorris commented:
Ready for re-review, @damonbayer

damonbayer commented on docs/source/tutorials/time.qmd:
```diff
@@ -10,7 +10,9 @@ The fundamental time unit should represent a period of fixed (or approximately f
 For many infectious disease renewal models of interest, the fundamental time unit will be days, and we will proceed with this tutorial treating days as our fundamental unit.
 
-`pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Return values from `RandomVariable.sample()` are `tuples` or `namedtuple`s of `SampledValue` objects. Each such `SampledValue` is optiona ...
+`pyrenew` deals with time having `RandomVariable`s carry information about (i) their own time unit expressed relative to the fundamental unit (`t_unit`) and (ii) the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units. Return values from `RandomVariable.sample()` are `tuples` or `namedtuple`s of `SampledValue` objects. `SampledValue` objects can have `t_ ...
```
```suggestion
`pyrenew` deals with time by having `RandomVariable`s carry information about

1. their own time unit expressed relative to the fundamental unit (`t_unit`) and
2. the starting time, `t_start`, measured relative to `t = 0` in model time in fundamental time units.

Return values from `RandomVariable.sample()` are `tuples` or `namedtuple`s of `SampledValue` objects. `SampledValue` objects can have `t_start` and `t_unit` attributes.
```
